scipy/stats/_hypotests.py

```diff
@@ -663,3 +663,274 @@ def somersd(x, y=None):
+        idx = np.abs(TX) >= abs(TX_obs)
+    elif alternative == "less":
+        idx = TX <= TX_obs
+    elif alternative == "greater":
+        idx = TX >= TX_obs
+    else:
+        msg = (
+            "`alternative` should be one of {'two-sided', 'less', 'greater'},"
+            f" found {alternative!r}"
+        )
+        raise ValueError(msg)
+
+    p_value = _binomial_maximisation_of_p_value_with_nuisance_param(
+        total_c1, total_c2, idx, num_it
+    )
+    return BarnardExactResult(TX_obs, p_value)
+
+
+def _binomial_maximisation_of_p_value_with_nuisance_param(
+    total_c1, total_c2, idx, num_it
+):
+    """
+    Maximisation of the pvalue in respect of a nuisance parameter considering
+    a 2x2 sample space.
+    """
+    n = total_c1 + total_c2
+    x1 = np.arange(total_c1 + 1, dtype=np.int64).reshape(-1, 1, 1)
+    x2 = np.arange(total_c2 + 1, dtype=np.int64).reshape(1, -1, 1)
+
+    x1_comb = _compute_log_combinations(total_c1)
+    x2_comb = _compute_log_combinations(total_c2)
+
+    nuisance_num = 100
+    inf_bound, sup_bound = 0, 1
+
+    for _ in range(num_it):
+        nuisance_arr = np.linspace(
+            start=inf_bound, stop=sup_bound, num=nuisance_num
+        )
+        # Reshape in dimension 3 array
+        nuisance_arr = nuisance_arr.reshape(1, 1, -1)
+
+        with np.errstate(divide="ignore", invalid="ignore"):
+            log_nuisance_arr = np.log(
+                nuisance_arr,
+                out=np.zeros_like(nuisance_arr),
+                where=nuisance_arr >= 0,
+            )
+            log_1_minus_nuisance_arr = np.log(
+                1 - nuisance_arr,
+                out=np.zeros_like(nuisance_arr),
+                where=1 - nuisance_arr >= 0,
+            )
+
+            nuisance_power_x1_x2 = log_nuisance_arr * (x1 + x2)
+            nuisance_power_x1_x2[(x1 + x2 == 0)[:, :, 0]] = 0
+
+            nuisance_power_n_minus_x1_x2 = log_1_minus_nuisance_arr * (
+                n - x1 - x2
+            )
+            nuisance_power_n_minus_x1_x2[(x1 + x2 == n)[:, :, 0]] = 0
+
+            PX = np.exp(
+                x1_comb[x1]
+                + x2_comb[x2]
+                + nuisance_power_x1_x2
+                + nuisance_power_n_minus_x1_x2
+            )
+
+        PX /= PX.sum(axis=(0, 1)).reshape(1, 1, -1)  # This operation is to
+        # compensate numerical errors because sums of PX should always be equal
+        # to one.
+
+        sum_PX = PX[idx].sum(axis=0)  # Just sum where TX >= TX0
+        max_nuisance_idx = sum_PX.argmax()
+
+        inf_bound = (
+            nuisance_arr[0, 0, max_nuisance_idx - 1]
+            if max_nuisance_idx > 0
+            else nuisance_arr[0, 0, 0]
+        )
+        sup_bound = (
+            nuisance_arr[0, 0, max_nuisance_idx + 1]
+            if max_nuisance_idx < nuisance_num - 1
+            else nuisance_arr[0, 0, -1]
+        )
+
+        p_value = sum_PX[max_nuisance_idx]  # take the max value
```

Comment:
This can be outside of the loop
```suggestion
    p_value = sum_PX[max_nuisance_idx]  # take the max value
```

Reply:
Change in c2e23644249e00ecc833f54f6517f4d2426f4779

Reply:
Thanks, update in c2e23644249e00ecc833f54f6517f4d2426f4779